test/bin/jar.spec.ts

```diff
@@ -1,26 +1,86 @@
-import jar from "keycloakify/bin/tools/jar";
-import { it, describe, vi } from "vitest";
-
-vi.mock("fs", () => ({ promises: { mkdir: () => {} }, createWriteStream: () => {} }));
-vi.mock("stream", async () => {
-    const readableMock = () => {
-        const mockDecorators = {
-            on: () => mockDecorators,
-            pipe: () => mockDecorators
-        };
-        return {
-            from: () => mockDecorators
-        };
-    };
-
-    return {
-        // @ts-ignore
-        ...(await vi.importActual("stream")),
-        Readable: readableMock()
-    };
-});
+import { jarStream, type ZipEntryGenerator } from "keycloakify/bin/tools/jar";
+import { fromBuffer, Entry, ZipFile } from "yauzl";
+import { it, describe, assert } from "vitest";
+import { Readable } from "stream";
+
+type AsyncIterable<T> = {
+    [Symbol.asyncIterator](): AsyncIterableIterator<T>;
+};
+
+async function arrayFromAsync<T>(asyncIterable: AsyncIterable<T>) {
+    const chunks: T[] = [];
+    for await (const chunk of asyncIterable) chunks.push(chunk);
+    return chunks;
+}
+
+async function readToBuffer(stream: NodeJS.ReadableStream) {
+    return Buffer.concat(await arrayFromAsync(stream as AsyncIterable<Buffer>));
+}
+
+function unzip(buffer: Buffer) {
+    return new Promise<ZipFile>((resolve, reject) =>
+        fromBuffer(buffer, { lazyEntries: true }, (err, zipFile) => {
+            if (err !== null) {
+                reject(err);
+            } else {
+                resolve(zipFile);
+            }
+        })
+    );
+}
+
+function readEntry(zipFile: ZipFile, entry: Entry): Promise<Readable> {
+    return new Promise<Readable>((resolve, reject) => {
+        zipFile.openReadStream(entry, (err, stream) => {
+            if (err !== null) {
+                reject(err);
+            } else {
+                resolve(stream);
+            }
+        });
+    });
+}
+
+function readAll(zipFile: ZipFile): Promise<Map<string, Buffer>> {
+    return new Promise<Map<string, Buffer>>((resolve, reject) => {
+        const entries1: Map<string, Buffer> = new Map();
+        zipFile.on("entry", async (entry: Entry) => {
+            const stream = await readEntry(zipFile, entry);
+            const buffer = await readToBuffer(stream);
+            entries1.set(entry.fileName, buffer);
+            zipFile.readEntry();
+        });
+        zipFile.on("end", () => resolve(entries1));
+        zipFile.on("error", e => reject(e));
+        zipFile.readEntry();
+    });
+}
+
 describe("jar", () => {
-    it("creates jar artifacts without error", () => {
-        jar({ artifactId: "artifactId", groupId: "groupId", rootPath: "rootPath", targetPath: "targetPath", version: "1.0.0" });
+    it("creates jar artifacts without error", async () => {
```

Comment:
I'm a little on the fence about this test; on the one side, it uses `Buffer` instead of reading from the file system, so it won't cover the use exactly. In my repos, I'd go forth and create a temp dir, not sure what implications that might have for the CI, so I deferred that. @garronej if you agree, I'll add another test which works with actual files instead of testing in-memory

Reply:
Of course, you have my approval for that.